init-tools.sh

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+
+__scriptpath="$(cd "$(dirname "$0")"; pwd -P)"
+__init_tools_log=$__scriptpath/init-tools.log
+__PACKAGES_DIR=$__scriptpath/.nuget/packages
+__TOOLRUNTIME_DIR=$__scriptpath/build_tools
+__DOTNET_PATH=$__scriptpath/.dotnet_stage0
+__DOTNET_CMD=$__DOTNET_PATH/dotnet
+if [ -z "$__BUILDTOOLS_SOURCE" ]; then __BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json; fi
+__BUILD_TOOLS_PACKAGE_VERSION=$(cat $__scriptpath/BuildToolsVersion.txt)
+__BUILD_TOOLS_PATH=$__PACKAGES_DIR/Microsoft.DotNet.BuildTools/$__BUILD_TOOLS_PACKAGE_VERSION/lib
+__PROJECT_JSON_PATH=$__TOOLRUNTIME_DIR/$__BUILD_TOOLS_PACKAGE_VERSION
+__PROJECT_JSON_FILE=$__PROJECT_JSON_PATH/project.json
+__PROJECT_JSON_CONTENTS="{ \"dependencies\": { \"Microsoft.DotNet.BuildTools\": \"$__BUILD_TOOLS_PACKAGE_VERSION\" }, \"frameworks\": { \"netcoreapp1.0\": { } } }"
+__CHANNEL=$(awk -F "=" '/CHANNEL/ {print $2}' "$__scriptpath/branchinfo.txt")
+
+# Increases the file descriptors limit for this bash. It prevents an issue we were hitting during restore
+FILE_DESCRIPTOR_LIMIT=$( ulimit -n )
+if [ $FILE_DESCRIPTOR_LIMIT -lt 1024 ]
+then
+    echo "Increasing file description limit to 1024"
+    ulimit -n 1024
+fi
+
+# Disable first run since we want to control all package sources
+export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+
+if [ ! -e "$__PROJECT_JSON_FILE" ]; then
+    "$__scriptpath/scripts/obtain/dotnet-install.sh" --install-dir "$__DOTNET_PATH" --channel $__CHANNEL
+
+    if [ ! -d "$__PROJECT_JSON_PATH" ]; then mkdir -p "$__PROJECT_JSON_PATH"; fi
+    echo $__PROJECT_JSON_CONTENTS > "$__PROJECT_JSON_FILE"
+
+    if [ ! -e "$__BUILD_TOOLS_PATH" ]; then
+        echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
+        echo "Running: $__DOTNET_CMD restore \"$__PROJECT_JSON_FILE\" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE" >> "$__init_tools_log"
+        "$__DOTNET_CMD" restore "$__PROJECT_JSON_FILE" --packages "$__PACKAGES_DIR" --source "$__BUILDTOOLS_SOURCE" >> "$__init_tools_log"
+        if [ ! -e "$__BUILD_TOOLS_PATH/init-tools.sh" ]; then echo "ERROR: Could not restore build tools correctly. See '$__init_tools_log' for more details."; fi
+    fi
+
+    echo "Initializing BuildTools..."
+    echo "Running: $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR" >> "$__init_tools_log"
+    "$__BUILD_TOOLS_PATH/init-tools.sh" "$__scriptpath" "$__DOTNET_CMD" "$__TOOLRUNTIME_DIR" >> "$__init_tools_log"
+    if [ "$?" != "0" ]; then
+        echo "ERROR: An error occured when trying to initialize the tools. Please check '$__init_tools_log' for more details."
```

Comment:
Should StdErr be redirected as well in all of these above commands if we expect error details to be in the log file?